### ISSUE TITLE
Gradle Play Plugin: Issues / Wishlist (based on 3.1.0-078266b8-SNAPSHOT)

### ISSUE THREAD

**Issue by Nezisi** [playframework/playframework#12720]
Hello,

My company is currently testing a possible migration of ~ 20 SBT projects to the new Gradle play plugin.

Would love to submit patches / help out if possible, thus the ticket and not an answer directly to the discussion https://github.com/orgs/playframework/discussions/12338 .

Task was recently assigned to me, given that the last update in the discussion thread was in March and I don't see a Play 3.1 branch yet, I hope I'm not too late to join the party ;)

To minimize “ticket bureaucracy”, I would write more specific tickets and close this ticket based on your feedback / suggestions.

All issues / wishlist are mostly nuisances, though some are problematic regarding user experience / security.

#### Description
- https://github.com/playframework/playframework/blob/a27ed61a234e2603a803e56a4a385ea6a2457229/dev-mode/gradle-plugin/src/main/java/play/gradle/plugin/PlayPlugin.java#L72-L81
- PlayPlugin -> Twirl Imports is added without ability of customization
- https://github.com/playframework/playframework/blob/a27ed61a234e2603a803e56a4a385ea6a2457229/dev-mode/play-build-link/src/main/java/play/TemplateImports.java#L35-L38
- requires explicitly adding the PlayImport.javaForms component
- https://github.com/playframework/playframework/blob/a27ed61a234e2603a803e56a4a385ea6a2457229/dev-mode/sbt-plugin/src/main/scala/play/sbt/PlayImport.scala#L45

#### Suggestion
- Configuration should be possible, to prevent adding unnecessary dependencies, which increases attack surface

#### Description
- for example 
`[Warn] : javac: [path] Unexpected file on path: /home/build/.gradle/caches/modules-2/files-2.1/com.fasterxml.jackson.module/jackson-module-parameter-names/2.16.1/3366aba5a30f1d8a703c80016001d7f147a82cd8/jackson-module-parameter-names-2.16.1.pom
[Warn] : javac: [path] Unexpected file on path: /home/build/.gradle/caches/modules-2/files-2.1/com.fasterxml.jackson.module/jackson-module-scala_2.13/2.16.1/c329e104a146dcf390d372cdeecebca521a4da90/jackson-module-scala_2.13-2.16.1.pom`
- warning occurs on compileScala step
- I'd assume it is an upstream bug in Gradle -> Scala Compile, as a POM file doesn't belong on the compile classpath

#### Suggestion
- Validation that Play Gradle Plugin does not modify Scala Compile classpath and lacks a filter for POM files
- opening a bug in Gradle (?)
- warning is harmless, but probably confusing to users

#### Description
- https://github.com/playframework/playframework/blob/a27ed61a234e2603a803e56a4a385ea6a2457229/dev-mode/gradle-plugin/src/main/java/play/gradle/plugin/PlayPlugin.java#L92-L103
- Play layout is assumed by default, without the possibility to override
- Source sets are added without validation that directories exist 

#### Suggestion
- Boolean option to disable App Layout 
- though not strictly necessary, better check directories exist before unnecessary modifying configuration

#### Description
- https://github.com/playframework/playframework/blob/a27ed61a234e2603a803e56a4a385ea6a2457229/dev-mode/gradle-plugin/src/main/java/play/gradle/plugin/PlayPlugin.java#L105-L109
- note: as I struggled with the proper set up of Scala vs Twirl vs Java ( https://github.com/playframework/twirl/issues/745 ), I can understand why this was added

#### Possible scenarios:
1) Project can have multiple source sets with different settings which are now "mingled together)
2) Scala compiler does - to my knowledge - not inherit Java compiler options, java compiler options have to be passed explicitly to the scala compiler via options.compilerArgs
-> thus the two different source sets could be compiled with entirely different compiler args
-> while code might compile successfully, it might lead to numerous runtime problems (think of Jackson "-parameter" flag to Javac for example or a missing "-release" flag to Javac for byte code compatibility)

#### Suggestion
- leave the source sets as is
- add a warning to Gradle if scala sourceset is empty and java sourceset is not empty (?)

#### Description
https://github.com/orgs/playframework/discussions/12338#discussioncomment-8943449

#### Solution
- adding the remaining parameters as optionals to PlayRunParams, with current default value 

If I have forgotten something, feel free to add it.

@ihostage Would be great to hear your feedback on this, please give me a ping if you agree with my solutions or have any questions.
Based on your feedback, I'd open a separate PR for each issue ;)

**Comment by mkurz** [comment]
> I don't see a Play 3.1 branch yet, I hope I'm not too late to join the party ;)

No, you are not too late ;) Pull requests regarding Gradle would definitely be considered for Play's next major release.

I did not implement the Gradle plugin, but still I allow myself to add some comments:

- Point 1 (imports):
  Having all the imports added automatically is the case when using sbt as well, so this just resembles what's happening when using sbt.
  Or did you customize the imports in your sbt project's? That is possible, so I am just interested if you did that and why you came up with this point.
  Also what do you mean with "_requires explicitly adding the PlayImport.javaForms component_"?" I mean I now what javaforms is, but what do you mean by it in the context of Gradle? (Again, I am not totally familiar with the Gradle plugin yet).
  
- Point 2 (javac warning)
  Not sure why this is happening. Like said I do not have lots of knowledge about the Gradle plugin at this point, so we either have to filter out pom files and/or yes, have to open a bug report at the Gradle repo.
  
- Point 3 (app layout)
  Agree, would be good to make layout configurable. IMHO detecting directories is overkill and might result in undesired behviour, I think it's better to have an explicit setting without any magic.
  Can you provide a pull request for such a setting?
  
- Point 4 (source sets)
  I leave this to @ihostage.

- Point 5 (PlayRunParams)
  Sorry I don't follow. Which default values do you mean? The discussion you refer is about jvm flags.

**Comment by ihostage** [comment]
Hi, @Nezisi! Thank you a lot for your feedback 👍 

> To minimize “ticket bureaucracy”, I would write more specific tickets and close this ticket based on your feedback / suggestions.

I don't see a problem split this issue to several and discuss their separately before make a pull requests 😉 
@mkurz don't you mind?

**Comment by mkurz** [comment]
> @mkurz don't you mind?

Of course not, please go ahead everyone.

**Comment by ihostage** [comment]
Hi, @Nezisi! 👋 
I decided write some comments here while you don't create a separated issues.

#### 1. Security (as adding unnecessary dependencies without possibility of configuration)
Fixed in #12771 

#### 2. Javac: Unsupported file on classpath
Can you create a small example to reproduce this warning?

#### 3. Play App Layout assumed by default, without the ability to override
I think we shouldn't do anything and both ([Play](https://www.playframework.com/documentation/3.0.x/Anatomy#The-Play-application-layout) and [Gradle/Sbt](https://www.playframework.com/documentation/3.0.x/Anatomy#Default-sbt-layout)) layouts should be supported by default without special configuration like now. If someone have a special layout, it can configure it by Gradle like any other source set. What the real problem is? Maybe I miss something 🤔 

#### 4. Java source set is explicitly converted to Scala
While Play Route compiles to Scala sources I don't see another options 🤷‍♂️ Yes, @mkurz already know, that my dream is Play Java users shouldn't contact with scalac (!= scala-library) if they don't use a Twirl. But currently it's impossible. The [Gradle Play Plugin](https://github.com/gradle/playframework/) did the [same things](https://github.com/gradle/playframework/blob/818d998bcca6bce1d2e8d7deaaecaf7a19be55a1/src/main/java/org/gradle/playframework/plugins/PlayApplicationPlugin.java#L99-L114). If you create a small project to present your pain and describe how you see a good behaviour Play Gradle Plugin we will discuss it of course.

#### 5. PlayRunParams should be extended to cover the currently hardcoded default values
This case is more complicated then looks one the first eye. Need to discuss what your real case is to understand how Play Gradle Plugin can help you.

**Comment by Nezisi** [comment]
Ticket can be closed. I will open a new ticket for two bugs happening in the Gradle plugin, most of the part above is now either outdated or irrelevant.
